portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -42,16 +44,43 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String name = getName(request);
     String comment = getComment(request);
+
     // Prevents users from submitting comments with no content."
     if (comment.equals("")) {
         response.setContentType("text/html");
         response.getWriter().println("Please enter a comment");
       return;
     }
-    comments.add(new Comment(comment, name));
+    // Store new comment into datastore
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(createCommentEntity(name, comment));
+
     response.sendRedirect("/index.html");
   }
 
+  // Retrieves comments from datastore and converts them into Comment objects.
+  private ArrayList<Comment> getComments() {
+    Query query = new Query("Comment");
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    ArrayList<Comment> comments = new ArrayList<Comment>();
+    for (Entity entity: results.asIterable()) {
+      String name = (String) entity.getProperty("sender");
+      String text = (String) entity.getProperty("text");
+      comments.add(new Comment(name, text));
```

Comment:
Thought exercise: What happens if there are too many comments to load into memory?

Reply:
Hmm that's interesting. I remember learning about memoization in javascript last semester, which is basically loading data in chunks and remembering them, and you can always retrieve more if you want without recalculating/reloading everything you've already loaded. I think memoization is more for speed optimization for expensive functions but maybe taking the idea of chunking comments together and handling those rather than all comments at once?